File: sem-2/Assignment/wholenumber_ADT_Assignment-1/1.cpp
#include <iostream>
#include <limits>

using namespace std;

typedef unsigned int wholeNum;

wholeNum get_zero() {
    return 0;
}

bool is_zero(wholeNum x) {
    return x == 0;
}

wholeNum add(wholeNum x, wholeNum y) {
    if (x > numeric_limits<wholeNum>::max() - y) {
        return numeric_limits<wholeNum>::max();
    } else {
        return x + y;
    }
}

wholeNum subtract(wholeNum x, wholeNum y) {
    if (x < y) {
        return 0;
    } else {
        return x - y;
    }
}

wholeNum divide(wholeNum x, wholeNum y) {
    if (y == 0) {
        return 0;
    } else {
        return x / y;
    }
}

wholeNum successor(wholeNum x) {
    return x + 1;
}

wholeNum predecessor(wholeNum x) {
    if (x == 0) {
        return 0;
    } else {
        return x - 1;
    }
}

bool equal(wholeNum x, wholeNum y) {
    return x == y;
}

int main() {
    cout << "true: 1; false: 0;" << endl;
    wholeNum x = 6;
    wholeNum y = 4;
    cout << "Get Zero: " << get_zero() << endl;
    cout << "Is Zero: " << is_zero(x) << endl;
    cout << "add (" << x << "," << y << "): " << add(x, y) << endl;
    cout << "subtract (" << x << "," << y << "): " << subtract(x, y) << endl;
    cout << "divide (" << x << "," << y << "): " << divide(x, y) << endl;
    cout << "successor (" << y << "): " << successor(y) << endl;
    cout << "predecessor (" << x << "): " << predecessor(x) << endl;
    cout << "equal (" << x << "," << y << "): " << equal(x, y) << endl;
    return 0;
}

File: sem-2/Assignment/wholenumber_ADT_Assignment-1/1.py
def get_zero():
    return 0

def is_zero(x):
    if x == 0:
        return True
    else:
        return False

def add(x, y):
    if x + y < 0:
        return 0
    elif x + y > pow(2, 31) - 1: 
        return pow(2, 31) - 1
    else:
        return x + y

def subtract(x, y):
    if x - y < 0:
        return 0
    elif x - y > pow(2, 31) - 1: 
        return pow(2, 31) - 1
    else:
        return x - y

def divide(x, y):
    if x <= 0 or y <= 0:
        return 0
    elif x // y > pow(2, 31) - 1: 
        return pow(2, 31) - 1
    else:
        return x // y

def successor(x):
    if x + 1 < 0:
        return 0
    elif x + 1 > pow(2, 31) - 1: 
        return pow(2, 31) - 1
    else:
        return x + 1

def predecessor(x):
    if x - 1 < 0:
        return 0
    elif x - 1 > pow(2, 31) - 1: 
        return pow(2, 31) - 1
    else:
        return x - 1

def equal(x, y):
    if x == y:
        return True
    else:
        return False

x = 6
y = 4
print("Get Zero:", get_zero())
print("Is Zero:", is_zero(x))
print("add (", x, ",", y, "):", add(x, y))
print("subtract (", x, ",", y, "):", subtract(x, y))
print("divide (", x, ",", y, "):", divide(x, y))
print("successor (", y, "):", successor(y))
print("predecessor (", x, "):", predecessor(x))
print("equal (", x, ",", y, "):", equal(x, y))
